Allow actions to be accessible from a web browser

rabbah opened the pull request:
Adds route to allow web based activations. Actions invoked this way are anonymous in that the
caller is not authenticated. The intended action must be named in the path as a fully qualified
name as in `/experimental/web/some-namespace/some-package/some-action`. The package is optional in that the action may be in the default package. In which case, the string "default" must be used.

If the action doesn't exist (or the namespace is not valid) a `BadRequest` is generated.
Optionally, the result form the action may be projected based on a named property. As in
`/experimental/web/some-namespace/some-package/some-action/some-property`. If the property
does not exist in the result then a `BadRequest` is generated. By convention, the "html" property will attempt to respond with media type "text/html".

Actions may be exposed to this web proxy by adding an annotation ("export" -> true).
Demo video https://ibm.box.com/s/5c6ignvejihbai3f59uvqcxee9etf0lf.

Note the first commit rebases #1740. 
Fixes #1222.
Fixes #1449.

rabbah commented:
Suggestions from the dev list (credit: Felix Meschberger):

- As for the status codes, 404/Not Found for non-existing actions rather than 400/Bad Request.
- And 404 as well for the case of the action not being available anonymously.
- Missing support for the requested content type could be 406/Not Acceptable (or also 404/Not Found).

rabbah commented:
note to reviewers: I removed two commits which will come by way of #1720 so next travis will break pending a rebase.

markusthoemmes commented on common/scala/src/main/scala/whisk/core/entity/EntityPath.scala:
```diff
@@ -170,6 +170,7 @@ protected[core] class EntityName private (val name: String) extends AnyVal {
     def toJson = JsString(name)
     def toPath: EntityPath = EntityPath(name)
     def addPath(e: EntityName): EntityPath = toPath.addPath(e)
+    def addPath(e: Option[EntityName]): EntityPath = e map { toPath.addPath(_) } getOrElse toPath
```
Is this used that often, that it's worth hiding the `Option` handling?

rabbah replied:
No but it's going to exist somewhere (e.g. Meta.api), might as well put it here for convenience.

markusthoemmes commented on common/scala/src/main/scala/whisk/core/entity/Identity.scala:
```diff
@@ -37,27 +37,38 @@ object Identity extends MultipleReadersSingleWriterCache[Identity, DocInfo] with
+        implicit val ec = datastore.executionContext
+        val ns = namespace.asString
+
+        cacheLookup(ns, {
```
As noted in a private chat: Can this cause problems with the `cacheKeyForUpdate` above?

rabbah replied:
No because `cacheKeyForUpdate` is for _updates_ and this is a read which provides an explicit key to lookup.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -16,23 +16,135 @@
+        Try {
+            val JsString(str) = result
+            val input = new ByteArrayInputStream(str.getBytes)
+            scala.xml.XML.load(input)
```
Is it necessary to parse this? Note that html5 is not necessary valid XML and thus this might fail. Also this puts unnecessary load on the controller i'd say.

rabbah replied:
ooops - left over from early debugging. will nuke.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -16,23 +16,135 @@
+            case JsString(s)  => complete(OK, s)
+            case JsBoolean(b) => complete(OK, b.toString)
+            case JsNumber(n)  => complete(OK, n.toString)
+            case JsNull       => complete(OK)
```
Are those even valid JSON responses? JSON only allows object or array as the top-level type.

rabbah replied:
A proper JSON response is indeed either an array or object, not a primitive type.
So the marshaler could treat these as error cases instead and if the caller desired to serialize primitive values then they must request `text` as the extension.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -16,23 +16,135 @@
+                        } else throw new Throwable("Illegal code")
+
+                    case _ => throw new Throwable("Illegal code")
+                } getOrElse 200
```
You might want to check out https://github.com/spray/spray/blob/master/spray-http/src/main/scala/spray/http/StatusCode.scala#L29-L35 which also throws an `Exception` on an illegal code and does all the standard handling for you.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -16,23 +16,135 @@
+
+    val supportedMediaTypes = mediaTranscoders.keySet
+
+    def metadata(method: String, headers: String, path: String, namespace: Option[String]) = {
```
Why is `headers` a `String` vs. a `List[String]`?

rabbah replied:
Fair question. I was avoid two mkstrings. You'll notice on like 54 there's a `headers.mkString("\n")` which will become `headers.map(_.toString)` and then on line 77 the code changes to `headers.mkString("\n").toJson`. I'm not opposed to changing it, it's probably cleaner. The reason `metadata` is a public method is to expose it to the test, so the added typing is arguably less important.

markusthoemmes replied:
Yes, the `headers.mkString` feels quite ugly though. I'd +1 for changing it.

rabbah replied:
OK - then might as well pass the headers as an array of strings to the action.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -44,27 +156,35 @@ trait WhiskMetaApi extends Directives with PostActionActivation {
 
-    /** Extracts the HTTP method, query params and unmatched (remaining) path. */
+    private lazy val validNameSegment = pathPrefix(EntityName.REGEX.r)
+    private lazy val packagePrefix = pathPrefix("""\Adefault\z""".r | EntityName.REGEX.r)
```
I believe `"default".r` is sufficient and equivalent to `"""\Adefault\z""".r` as regexes are anchored in Scala by default.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -16,23 +16,135 @@
+    val mediaTranscoders = Map(
+        "html" -> resultAsHtml _,
+        "text" -> resultAsText _,
+        "js" -> resultAsJson _,
```
Why use a non-standard `js` vs `json`? Note that they are *not* the same (JSON is way more strict than pure JavaScript notation). I'd change this to json.

rabbah replied:
hence the relaxed marshaling of json values.

markusthoemmes replied:
Don't know how to feel about both to be honest. Feels a bit unclean and non-standard. Do I read it correctly though, that `string`, `boolean` and `number` will be sent as `text/plain` vs objects and arrays being sent as `application/json`?

rabbah replied:
That's the correct interpretation of the code as it is.  @sjfink want to further justify the `js` choice vs `json`?

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -19,20 +19,127 @@ package whisk.core.controller
+                } getOrElse (OK)
+
+                respondWithHeaders(headers) {
+                    complete(code, JsObject())
```
Why not also return a `body` field of the JSON result as a `String` here? The user could then set `Content-Type` via the headers, effectively giving him full control over the HTTP response.

rabbah replied:
:) Yes, I know. I got lazy.

rabbah replied:
:) Yes, I know. I got lazy. You shall have it and it will be beautiful.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -44,27 +151,35 @@ trait WhiskMetaApi extends Directives with PostActionActivation {
-    /** Extracts the HTTP method, query params and unmatched (remaining) path. */
+    private lazy val validNameSegment = pathPrefix(EntityName.REGEX.r)
+    private lazy val packagePrefix = pathPrefix("default".r | EntityName.REGEX.r)
+    private lazy val extensionSplitter = EntityName.REGEX.dropRight(2).concat("""\.([a-zA-Z0-9.+-]+)\z""").r
```
As the `anchored` comment from above seems to be true, should we alter `EntityName.REGEX` to be `"""([\w]|[\w][\w@ .-]*[\w@.-]+)"""` thus removing the nasty `dropRight(2)` and end up with: s"""${EntityName.REGEX\.([a-zA-Z0-9.+-]+)""".r`?

rabbah replied:
Rather not mess with it because I'll have to think about the implications of that (i.e., where are all the places we use that regex match). I think being explicit does not hurt (i.e., specifying the anchors and make it clear this is a full match). I should add a comment to explain the drop right though in that case.

rabbah replied:
As expected - removing the anchors from EntityName.REGEX fails all the unit tests...

rabbah replied:
OK now I remember - If look though the implementation of `pathPrefix` matching for a regex. It uses `findPrefixMatchOf` which will not match the entire segment.

rabbah replied:
The regex is `([\w]|[\w][\w@ .-]*[\w@.-]+)` but perhaps should be `([\w]+|[\w][\w@ .-]*[\w@.-]+)` instead.

markusthoemmes replied:
that means my comment on `"""default""".r` was wrong as well I suppose.

rabbah replied:
No - it means we can change the EntityName.REGEX. Since "default" is a complete match it's OK.

rabbah replied:
And your observations shows there's a bug in the meta route matcher.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -19,20 +19,127 @@ package whisk.core.controller
+
+    def metadata(method: String, headers: List[String], path: String, namespace: Option[String]) = {
+        Map("__ow_meta_verb" -> method.toLowerCase.toJson,
+            "__ow_meta_headers" -> headers.toJson,
```
I'm still wondering: What will this array look like now? Something like

```
["Authentication: Basic Foo", "Custom-Header": "bar"]
```
?

Would make sense to even pass the headers as a `Map` to make header handling closer to what *express* and other frameworks do in node.js for example?

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -19,20 +19,127 @@ package whisk.core.controller
+
+    val supportedMediaTypes = mediaTranscoders.keySet
+
+    def metadata(method: String, headers: List[String], path: String, namespace: Option[String]) = {
```
I might well be missing something, but would it make sense to have all those types be the strongest possible and  do all the `toString` magic in this method? I'd rather push working with `String` as far out as possible.

psuter commented:
Some high-level comments from a first, superficial, pass:

* `"export": true` is a very generic-sounding annotation. Maybe `web_export`? (do we CamelCase?)
* I'm a concerned with the monumental work that comes with trying to encode an HTTP request into JSON fields (passed to the action), and interpreting a JSON object as an HTTP response. I'm afraid that conventions can only ever get us the easy cases, and that soon enough we'll be maintaining and documenting dozens of them. Could we instead resort to some externally documented schema? E.g. HAR has subsections for [requests](http://www.softwareishard.com/blog/har-12-spec/#request) and [responses](http://www.softwareishard.com/blog/har-12-spec/#response), but they may be other formats. Within actions, I'd expect that we will rapidly support translating these encoded objects into user-friendly libraries (e.g. `request`).
* With the comment above in mind, I would be in favor of simplifying the implementation and removing path projection (the path would be one of the components passed to the action, which can decide how to handle it).

rabbah commented:
It's fair to change the annotation. When this is moved out of experimental the annotation should then be promoted to a first class property which might require a schema change.

Note that the http bits are the last commit. Much of the work in place is to enable the web based activation. For that and for making standard actions accessible without buying into http request/response, the projection is still useful. That said as you noticed the action does receive the unmatched parts and can decide how to respond anyway. This I posit makes for a gentler migration for actions. For moving express apps the counter is likely true and your points are well taken. 

For the http bits what I supported is header setting and nothing else. This was motivated by a use case around oauth flow where having redirects is helpful.

Your suggestion of not reinventing the wheel and using some accepted schema based approach is all goodness. I'm for it.

rabbah commented:
@csantanapr any thoughts on this pr?
@markusthoemmes @psuter I think I addressed the actionable issues from your first pass. I would like to defer the http schema handling to a future enhancement.

rabbah commented:
PG3/176 approved.

rabbah commented:
PG3/180 approved.

csantanapr commented:
@rabbah What would be the problem to use `.json` instead of `.js` in the web api url `.../action.json` ?
I think using `.json` is more clear to talk about `application/json` instead of `.js` or maybe I'm missing something.

rabbah commented:
Great comments @csantanapr. We can add CORS support - it's in the controller already, just have to wrap these new routes with it -- I think at cursory glance. Yes images could be supported. I opted not to include it here but did try it (with an svg). I agree on POST but adding it without supporting form-urlencoded input will be broken, so we'll have to add both at once. Lastly, it's .js not .json deliberately. This is raised by Markus also. It would make projecting a JSON value (string) for example an error (one must use a text extension - similarly for the other value types). So it's a convenience so that if one is projecting top level fields or nested fields they're not constantly changing both the extension and the projection at the same time. I can be swayed toward stricter form though. I'd like to consider all but the last as additive features to address as future enhancements.

csantanapr commented:
@rabbah My comment was just feedback for additive future enhancements to discuss, and possible to break into different issues/PRs .

csantanapr commented:
No magic 🌮 🌮

csantanapr commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -448,6 +449,10 @@ trait WhiskMetaApi extends Directives with PostActionActivation {
                 // blocking invoke which got queued instead
                 complete(Accepted, JsObject("code" -> transid.id.toJson))
 
+            case Failure(t: BlockingInvokeTimeout) =>
```
This will give us the tid for api-experimental meta calls that go beyond the limit timeout. 👍

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -34,13 +38,110 @@ import whisk.core.entity.types._
+    def withBody(b: Option[JsObject]) = Context(method, headers, path, query, b)
+    def metadata(user: Option[Identity]): Map[String, JsValue] = {
+        Map("__ow_meta_verb" -> method.value.toLowerCase.toJson,
+            "__ow_meta_headers" -> headers.map(h => h.name -> h.value).toMap.toJson,
```
Should we normalize the `name` to be lowercased as well? There is `lowercasedName` on `HttpHeader`

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -34,13 +38,110 @@ import whisk.core.entity.types._
+        } match {
+            case Success(xml) => respondWithMediaType(`text/html`) { complete(OK, xml) }
+            case Failure(t)   => complete(BadRequest, Messages.invalidMedia(`text/html`))
+        }
```
Might as well be just a `match`?

```
result match {
    case JsString(str) => respondWithMediaType(`text/html`) { complete(OK, str) }
    case _ => complete(BadRequest, Messages.invalidMedia(`text/html`))
}
```

rabbah replied:
left over from debugging/parsing xml.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -190,77 +448,108 @@ trait WhiskMetaApi extends Directives with PostActionActivation {
+            } getOrElse {
+                info(this, s"${action.fullyQualifiedName(true)} not exported")
+                Future.failed(RejectRequest(NotFound))
+            }
```
Minor scalaism nit: You can convert an `Option` to a `Boolean` by using `exists`:

```
if(action.annotations.asBool("web-export").exists(identity)) {
    info(this, s"${action.fullyQualifiedName(true)} is exported")
    Future.successful(action)
} else {
    info(this, s"${action.fullyQualifiedName(true)} not exported")
    Future.failed(RejectRequest(NotFound))
}
```

You decide what you like more.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -190,77 +448,108 @@ trait WhiskMetaApi extends Directives with PostActionActivation {
+                    isMetaPackage match {
+                        case Some(true) => Future.failed(RejectRequest(MethodNotAllowed))
+                        case _          => Future.failed(RejectRequest(NotFound))
+                    }
```
Same as below on converting `Option[Boolean]` to `Boolean`, handling `None` as `false`:

```
val isMetaPackage = pkg.annotations.asBool("meta").exists(identity)
if(isMetaPackage) {
    pkg.annotations.asString(method.name.toLowerCase).map { actionName => 
        info(this, s"'${pkg.name}' maps '${method.name}' to action '${actionName}'")
        Future.successful(EntityName(actionName), pkg.parameters)
    }.getOrElse {
        info(this, s"'${pkg.name}' is missing action name for '${method.name.toLowerCase}'")
        Future.failed(RejectRequest(MethodNotAllowed))
    }
} else {
    info(this, s"'${pkg.name}' is missing 'meta' annotation")
    Future.failed(RejectRequest(NotFound))
}
```

To me that is easier to follow what exactly is happening and as a bonus, we get a disambiguated log message.

rabbah replied:
thank you for the suggestion.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -115,67 +224,216 @@ trait WhiskMetaApi extends Directives with PostActionActivation {
+            context.requestParams -- action.parameters.immutableParameters
+        } else context.requestParams // in the absence of immutable annotations, return the request untouched
+
+        if (invokeParams.size == context.requestParams.size) {
```
Is there a reason you used the `intersect` method proposed by me above, but not in this case?

rabbah replied:
as we discussed - the reason was tied to where the immutable parameters are checked against the incoming parameters... the original thought was not to leak this to WhiskAction but that was already happening with the presence of the annotation check there anyway. Thanks for the patch that improves it.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Meta.scala:
```diff
@@ -527,22 +522,20 @@ trait WhiskMetaApi extends Directives with PostActionActivation {
-
-                case _ =>
+                } getOrElse {
                     info(this, s"'${pkg.name}' is missing 'meta' annotation or action name for '${method.name.toLowerCase}'")
```
That log line should be 

```
"'${pkg.name}' is missing action name for '${method.name.toLowerCase}'"
```

rabbah replied:
amended.

rabbah commented:
Rebased and squashed.

rabbah commented:
Rebased again to pick up the library bumps. Also added a CLI test. Removing blocked tag because this PR can be merged as is, or if staged with #1740, rebasing there after will remove the over lapping commits. The last commit in this pr should be squashed into the one that precedes it but is left on its own for the reviewer's benefit. PG1/1106 reapproved with latest test.

csantanapr commented:
#1740 is ready to be merge or your waiting on more reviews?

rabbah commented:
It's ready pending new input from @markusthoemmes.

sjfink commented:
pushed in b5be59f..09471ee
